docs/_ext/pharmpy_snippet.py

```diff
@@ -0,0 +1,199 @@
+from docutils import nodes
+from docutils.statemachine import ViewList, string2lines
+from docutils.parsers.rst import Directive, directives
+
+from conversion import transpile_py_to_r
+
+def setup(app):
+    app.add_directive('pharmpy-execute', PharmpyExecute)
+    app.add_directive('pharmpy-code', PharmpyCode)
+
+    return {
+        'version': '0.1',
+        'parallel_read_safe': True,
+        'parallel_write_safe': True,
+    }
+
+def csv_option(s):
+    return [p.strip() for p in s.split(",")] if s else []
+
+class RecursiveDirective(Directive):
+
+    def _convert_lines_to_nodes(self, lines: List[str]) -> List[nodes.Node]:
+        """Turn an RST string into a node that can be used in the document.
+
+           See https://github.com/sphinx-doc/sphinx/issues/8039
+        """
+
+        node = nodes.Element()
+        node.document = self.state.document
+        self.state.nested_parse(
+            ViewList(
+                string2lines('\n'.join(lines)),
+                source='[SnippetDirective]',
+            ),
+            self.content_offset,
+            node,
+        )
+
+        return node.children
+
+        # for idx, line in enumerate(new_content):
+            # self.content.data.insert(idx, line)
+            # self.content.items.insert(idx, (None, idx))
+
+        # del self.content.data[len(new_content):]
+        # del self.content.items[len(new_content):]
+
+        # node = nodes.container()
+        # self.state.nested_parse(self.content, self.content_offset, node)
+
+class PharmpyExecute(RecursiveDirective):
+    required_arguments = 0
+    optional_arguments = 0
+    final_argument_whitespace = True
+    has_content = True
+
+    option_spec = {
+        "hide-code": directives.flag,
+        "hide-output": directives.flag,
+        "code-below": directives.flag,
+        "linenos": directives.flag,
+        "lineno-start": directives.nonnegative_int,
+        "emphasize-lines": directives.unchanged_required,
+        "raises": csv_option,
+        "stderr": directives.flag,
+    }
+
+    def run(self):
+        # if 'hide-code' in self.options:
+            # raise ValueError('pharmpy-execute does not allow to hide code, use jupyter-execute instead')
+
+        return self._nodes()
+
+    def _nodes(self):
+        lines = self._lines()
+        return self._convert_lines_to_nodes(lines)
+
+    def _lines(self):
+        return [
+            f'.. container:: pharmpy-snippet{"" if "hide-output" in self.options else " with-output"}',
+            '',
+            *_indent(3, self._input_output_lines())
+        ]
+
+    def _input_output_lines(self):
+        if 'hide-code' in self.options:
+            return self._output()
+
+        if 'code-below' in self.options:
+            return [
+                *self._output(),
+                '',
+                *self._input(),
+            ]
+
+        return [
+            *self._input(),
+            '',
+            *self._output(),
+        ]
```

Comment:
We always have to return the output lines even when `hide-output` is set, otherwise the code will not be executed.

Reply:
Yes!